### ISSUE TITLE
proxy_temp_path и use_temp_path

### ISSUE THREAD

**Issue by RussianCatYakov** [webserver-llc/angie#70]
Привет!
Столкнулись с проблемой, что use_temp_path=off игнорируется в proxy_cache_path, кэш все равно продолжает писаться в темповую директорию, которая задана при билде.
Сборка на основе версии 1.1, темповая директория задана так
`--http-proxy-temp-path=/var/lib/angie/proxy 
`
Конфиг выглядит вот так
```
proxy_cache_path /dev/shm/cache_tmpfs keys_zone=cache_tmpfs:128m
    levels=1:2
    max_size=700g
    use_temp_path=off
    inactive=5m;

proxy_cache_path /opt/cache/cache1 keys_zone=cache1:4096m
    levels=1:2
    max_size=8000g
    use_temp_path=off
    inactive=10d;
```
По идее запись файлов кэша должна идти сразу в диск, но если аттачнуться стрейсом к одному из воркеров, то видно, что записи в темповую директорию идут.

```
$ sudo strace -p 2265415 2>&1 | grep var
openat(AT_FDCWD, "/var/lib/angie/proxy/3/93/1825445933", O_RDWR|O_CREAT|O_EXCL, 0600) = 297
unlink("/var/lib/angie/proxy/3/93/1825445933") = 0
openat(AT_FDCWD, "/var/lib/angie/proxy/5/52/1825446525", O_RDWR|O_CREAT|O_EXCL, 0600) = 391
unlink("/var/lib/angie/proxy/5/52/1825446525") = 0
openat(AT_FDCWD, "/var/lib/angie/proxy/2/20/1825447202", O_RDWR|O_CREAT|O_EXCL, 0600) = 297
unlink("/var/lib/angie/proxy/2/20/1825447202") = 0
openat(AT_FDCWD, "/var/lib/angie/proxy/6/73/1825447736", O_RDWR|O_CREAT|O_EXCL, 0600) = 300
unlink("/var/lib/angie/proxy/6/73/1825447736") = 0
openat(AT_FDCWD, "/var/lib/angie/proxy/6/47/1825448476", O_RDWR|O_CREAT|O_EXCL, 0600) = 383
unlink("/var/lib/angie/proxy/6/47/1825448476") = 0
openat(AT_FDCWD, "/var/lib/angie/proxy/8/11/1825449118", O_RDWR|O_CREAT|O_EXCL, 0600) = 393
unlink("/var/lib/angie/proxy/8/11/1825449118") = 0
```

**Comment by VBart** [comment]
Вы наблюдаете запись во временную директорию ответов, которые не подлежат кэшированию. По этой причине следом сразу следует вызов `unlink()`, который удаляет ссылку на файл из файловой системы.

Опция `use_temp_path` имеет действие только для кэшируемых ответов в рамках конкретной кэш-зоны.

**Comment by RussianCatYakov** [comment]
Не совсем понятно, что за файлы, которые не кэшируются. По нашей конфигурации даже не-200 коды ответа кэшируются. 
Данные операции создают очень неприятную нагрузку на диски.

**Comment by iHeadRu** [comment]
> Не совсем понятно, что за файлы, которые не кэшируются. По нашей конфигурации даже не-200 коды ответа кэшируются. Данные операции создают очень неприятную нагрузку на диски.

Вам намекают на proxy_buffering :)

**Comment by VBart** [comment]
Добавьте в access log логгирование переменной [`$upstream_cache_status`](https://angie.software/http_upstream/#upstream-cache-status) и будет видно. Причин, по которым ответ может не кэшироваться несколько и все они перечисленны в конце описания директивы [`proxy_cache_valid` ](https://angie.software/http_proxy/#proxy-cache-valid). Например, ответ может содержать заголовок `Set-Cookie`.

**Comment by VBart** [comment]
> У нас на некоторые локейшены стоит min_uses 2, срабатывает ли это правило на первый запрос - игнорируется use_temp_path и пишет в дефолтную временную директорию?

Да, всё так. Если запрос не кэшируется в том числе потому, что не достигнут `min_uses`, то ответ не будет сохраняться в кэш и в случае необходимости будет задействоваться временный файл в директории для временных файлов.

**Comment by RussianCatYakov** [comment]
Спасибо за помощь, теперь понятно. Максимально неочевидная штука, конечно :)

**Comment by RussianCatYakov** [comment]
Спасибо за помощь, теперь понятно. Максимально неочевидный случай :)
